### ISSUE TITLE
[class.mem] Chaging noexcept-specifier to noexcept operator in complete-class context

### ISSUE THREAD

**Issue by ranaanoop** [cplusplus/CWG#147]
Full name of submitter (unless configured in github; will be published with the issue): Anoop Rana

**Reference** (section label): [class.mem]

**Link to reflector thread (if any)**: https://stackoverflow.com/questions/73998553/why-is-that-noexcept-specifier-doesnt-work-in-complete-class-context-but-defaul

**Issue description**: Currently the complete-class context includes `noexcept-specifier` but I think that it should instead include `noexcept operator`. There are 2 important things to note here about noexcept operatror. First, the operand of an noexcept operator is a unevaluated operand unlike for noexcept-specifier. Second, the `noexccept-expression` of the noexcept operator is an integral constant expression whereas the `constant-expression` in a `noexcept-specifier` is a contextually converted constant expression of type `bool`. 

That is, with the `noexcept operator` the `noexcept-expression` can be of some other integral type like `int` etc and doesn't have to be just `bool`. So I think that this makes the `noexcept operator` to be a better candidate for a complete class context than `noexcept-specifier`. 

Also, as discussed in the linked [question](https://stackoverflow.com/questions/73998553/why-is-that-noexcept-specifier-doesnt-work-in-complete-class-context-but-defaul), even though `noexcept-specifier` is included in the complete class context, all compilers [reject](https://godbolt.org/z/njsqcTqhY) the following code(that uses noexcept-specifier):

```
struct A {
   constexpr static bool func() 
   { 
       return true; 
   }
   //--------------vvvvvv------->works as expected   #1
   void f(bool V1 = func())
   { 
      bool V2 = func(); //works as expected          #2
   }
   //-----------------vvvvvv---->REJECTED BY ALL COMPILERS       #3
   void g()  noexcept(func()) 
   {
      ;
   }

};
```

While all compilers accept the below code(that uses noexcept operator):
```
struct A {
   constexpr static int func() 
   { 
       return 5; 
   }
  
   void f(int V1 = func())
   { 
      int V2 = func(); 
   }
   //-----------------vvvvvvVVVVVVVVVV---->ACCEPTED BY ALL COMPILERS 
   void g()  noexcept(noexcept(func())) 
   {
      ;
   }

};
```

<h4>Suggested resolution:</h4>

Change [class.mem.general#7.4](https://eel.is/c++draft/class.mem.general#7.4) to as shown below:

> A [complete-class context](https://eel.is/c++draft/class.mem.general#def:complete-class_context) of a class (template) is a
>
> **7.4** noexcept operator([expr.unary.noexcept](https://eel.is/c++draft/expr.unary.noexcept))

**Comment by jensmaurer** [comment]
I believe the current state of the wording is intentional. What we have here is even easier than
```
struct S {
  constexpr static int size() { return 5; }
  int x[sizee()];   // not a complete class context
};
```
Please submit bug reports to vendors instead.

**Comment by ranaanoop** [comment]
@jensmaurer I actually did a lot of search and research before submitting this issue. For example, I also came across [this question] (https://stackoverflow.com/questions/29551223/static-constexpr-function-called-in-a-constant-expression-is-an-error) which basically has the same problem and if we see the [answer](https://stackoverflow.com/a/29662526/12002570) to that question, it mentions that the standard is not quite clear on this. They wrote in their [answer](https://stackoverflow.com/a/29662526/12002570)(maybe incorrectly:

> The central problem is that class members are generally not considered to be declared until after the class in which they're declared is complete. Thus, regardless of the fact that foo is static constexpr and its declaration precedes that of bar, it cannot be considered "available" for use in a constant expression until MyClass is complete.

(end quote)

As you suggested, I'll contact the vendors and report the bug. For now(for the time being), i would like that this issue be remained open and once I get any reply confirming that the compilers are at fault, I'll close this myself.

**Comment by jensmaurer** [comment]
I agree with the general quote, but I do note that we explicitly made _noexcept-specifiers_ complete-class contexts, clarifying that they are to be parsed as-if the class was complete at that point.

**Comment by frederick-vs-ja** [comment]
See also CWG1308 and CWG1330. I also believe that the status quo is intentional.

Your sugession would break the following example which is [currently accepted by all mainstream compilers](https://godbolt.org/z/PdKczvbvM) and clearly well-formed per current wording.
```C++
struct B {
   constexpr static bool func() noexcept(bool(sizeof(B))) // B must be complete in the exception specification
   { 
       return true; 
   }
};
```

According to the current wording, it's unclear to me why compilers consider that **the definition of `A::func` is unavailable at `#3`**. But it's clear to me that they think `A` itself is complete at `#3`, which is consistent with the standard wording.

This issue is possibly a duplicate of CWG2510 - I think member functions have the exactly same problem.

**Comment by ranaanoop** [comment]
@jensmaurer I submitted a [clang bug](https://github.com/llvm/llvm-project/issues/58245#issue-1402015438) and there we can see that one of their implementors(afaik) also noted that the version with the noexcept operator is accepted. Though they didn't specify the reason why the `noexcept-specifier` version is rejected by clang even though standard permits it.

**Comment by ranaanoop** [comment]
@frederick-vs-ja  The reason your example [works](https://godbolt.org/z/PdKczvbvM) is because just like in my [second example](https://godbolt.org/z/GKjb87n7Y), you're using another construct that has an unevaluated operand. That is, the only difference between my second example and your example is that I was using `noexcept operator` while you're using `sizeof`. And both of these constructs have an unevaluated operand. This is it works with all compilers. I am still not sure that if `noexcept-specifier` should be the list of complete-class context. 

@jensmaurer  I also came to know that "[Revision 4 of Declarations and where to find them](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1787r4.html) removed noexcept from the list of complete class contexts but it was added back in [revision 5 but no explanation in the revision history explain it](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1787r5.html)." 

I've added some more wording at the end of my current isuue(the suggested resolution part), saying this.

**Comment by ranaanoop** [comment]
@frederick-vs-ja  Now in my revised wording of the issue, i've suggested the following resolution. This also means that the [code](https://godbolt.org/z/PdKczvbvM) that you showed, will still work with the revised wording:

Either completely remove `noexcept-specifier` from the list of complete-class context or make it more specific by saying that "`noexcept-specifier` with constructs that take an unevaluated operand". 

This means,  change [class.mem.general#7.4](https://eel.is/c++draft/class.mem.general#7.4) to as shown below:

> A [complete-class context](https://eel.is/c++draft/class.mem.general#def:complete-class_context) of a class (template) is a
>
> **7.4** `noexcept-specifier` with constructs that take an unevaluated operand(like `sizeof`, noexcept opeator etc).

**Comment by jensmaurer** [comment]
>  one of their implementors(afaik) also noted that the version with the noexcept operator is accepted

When receiving a bug report for a compiler, it seems plausible to investigate the bounds of the non-conforming behavior. That, in itself, does not seem to constitute an opinion on whether the standard is defective in that area.

**Comment by jensmaurer** [comment]
I'm sorry, I'm not seeing a defect in the standard right now. The wording is sufficiently clear and very much intentional. I'm just seeing buggy implementations.
If the rules turn out to be unimplementable, CWG needs to have a look. No such argument has been made, yet. 

If you feel the rules should be changed in this area, please consider submitting a paper to WG21 with rationale.

**Comment by frederick-vs-ja** [comment]
> Either completely remove `noexcept-specifier` from the list of complete-class context or make it more specific by saying that "`noexcept-specifier` with constructs that take an unevaluated operand".

The first sugesstion is CWG1308, which has been reverted.

> This means, change [class.mem.general#7.4](https://eel.is/c++draft/class.mem.general#7.4) to as shown below:
> 
> > A [complete-class context](https://eel.is/c++draft/class.mem.general#def:complete-class_context) of a class (template) is a
> > **7.4** `noexcept-specifier` with constructs that take an unevaluated operand(like `sizeof`, noexcept opeator etc).

I don't think this suggestion should be accepted either. It would break the following currently well-formed example, which is also [accepted by mainstream compilers](https://godbolt.org/z/8qGj1njTY).
```C++
struct B1 {
    static void func() noexcept(B1::value) {} // B1::value requires B1 to be complete

    constexpr static bool value = true;
};
```

I think there are actually something unclear for the _noexcept-specifier_ (exception specification) and the function body of a member or friend functions, inside an in-class _noexcept-specifier_ of such a function.

But as CWG1330 required that the class type is complete in such a _noexcept-specifier_ and implementations agree with such decision, I don't think we should change the definition of complete-class context.

----
However, as CWG1308 has been reverted, is there anything forbidding this example?
```C++
struct Foo {
    static void func() noexcept(!noexcept(Foo::func())) {} // Circular dependency!
};
```
There is [implementation divergence](https://godbolt.org/z/Ynbn9a3P1): some reject it, some accept it with inconsistent semantics.

**Comment by ranaanoop** [comment]
@jensmaurer  Apparently, msvc team [says that](https://developercommunity.visualstudio.com/t/MSVC-rejects-use-of-static-constexpr-mem/10166701#TPIN-N10167796):

> Thanks for your reporting the issue to us. **But we determined that the issue is not a bug.**

As we can see in the above quoted statement(which is from my submitted bug report for msvc), msvc's team claim that this is not a bug and so they've closed the issue/bug that i reported/submitted  there.

**Comment by jensmaurer** [comment]
@frederick-vs-ja, I agree the circular case is something we need to make ill-formed; we don't want the sort of self-referential inconsistency you've shown. Given that a noexcept-specifier is a complete-class context, it seems multiple "hops" inside the class may be possible, so that the function is not actually directly mentioned in its noexcept-specifier.

@ranaanoop, that longer quote from MSVC people gives several good arguments why we might need to evaluate the noexcept-specifier of a member function while constructing the class body (e.g. because it can affect the type of a data member), something which complete-class contexts should not be able to do.

**Comment by ranaanoop** [comment]
Maybe the wording could be something like:

> While evaluating the result of a noexcept operator predicate nested inside a noexcept-specifier for a function F, the operand of the noexcept operand will not recursively refer F.

The same has add [here](https://github.com/cplusplus/CWG/issues/152#issuecomment-1278790478).
